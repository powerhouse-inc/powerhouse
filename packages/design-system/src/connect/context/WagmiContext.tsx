import { wagmiConfig as defaultWagmiConfig } from "@/services";
import {
  QueryClient,
  QueryClientProvider,
  QueryClientProviderProps,
} from "@tanstack/react-query";
import { ReactNode } from "react";
import { WagmiProvider, WagmiProviderProps } from "wagmi";

const defaultQueryClient = new QueryClient();

type Props = {
  readonly wagmiProviderProps?: WagmiProviderProps;
  readonly queryClientProviderProps?: QueryClientProviderProps;
  readonly children: ReactNode;
};
export function WagmiContext(props: Props) {
  const { children, wagmiProviderProps, queryClientProviderProps } = props;
  const { config = defaultWagmiConfig, ...wagmiProps } =
    wagmiProviderProps ?? {};
  const { client = defaultQueryClient, ...queryClientProps } =
    queryClientProviderProps ?? {};

  return (
    <WagmiProvider config={config} {...wagmiProps}>
      <QueryClientProvider client={client} {...queryClientProps}>
        {children}
      </QueryClientProvider>
    </WagmiProvider>
  );
}
