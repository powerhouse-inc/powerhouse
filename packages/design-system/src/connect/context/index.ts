export * from "./UiNodesContext";
export * from "./WagmiContext";
