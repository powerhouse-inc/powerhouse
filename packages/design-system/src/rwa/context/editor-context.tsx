import {
  CashAsset,
  EditorAction,
  EditorDispatcher,
  FixedIncome,
  getActionOperationType,
  getCashAsset,
  getFixedIncomeAssets,
  getStateKeyForTableName,
  getTableNameFor,
  makeTableData,
  Operation,
  RealWorldAssetsState,
  TableItemType,
  TableName,
} from "@/rwa";
import {
  createContext,
  ReactNode,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useReducer,
  useRef,
  useState,
} from "react";

export type RWAEditorContextProps = {
  readonly isAllowedToCreateDocuments: boolean;
  readonly isAllowedToEditDocuments: boolean;
  readonly canUndo: boolean;
  readonly canRedo: boolean;
  readonly onSwitchboardLinkClick: (() => void) | undefined;
  readonly onExport: () => void;
  readonly onClose: () => void;
  readonly onShowRevisionHistory: () => void;
};
type TEditorContext = RWAEditorContextProps &
  RealWorldAssetsState & {
    readonly cashAsset: CashAsset;
    readonly fixedIncomes: FixedIncome[];
    readonly selectedTableItem: TableItemType<TableName> | null;
    readonly selectedTableName: TableName | null;
    readonly operation: Operation;
    readonly viewItem: (
      item: TableItemType<TableName>,
      tableName: TableName,
    ) => void;
    readonly createItem: (tableName: TableName) => void;
    readonly editItem: (
      item: TableItemType<TableName>,
      tableName: TableName,
    ) => void;
    readonly clearSelected: () => void;
    readonly getIsFormOpen: (tableName: TableName) => boolean;
    readonly handleAction: (action: EditorAction) => void;
    readonly handleUndo: () => void;
    readonly handleRedo: () => void;
  };

const defaultEditorContextValue: TEditorContext = {
  accounts: [],
  fixedIncomeTypes: [],
  portfolio: [],
  principalLenderAccountId: "",
  serviceProviderFeeTypes: [],
  spvs: [],
  transactions: [],
  fixedIncomes: [],
  cashAsset: {
    id: "",
    type: "Cash",
    currency: "USD",
    balance: 0,
    spvId: "",
  },
  isAllowedToCreateDocuments: false,
  isAllowedToEditDocuments: false,
  canUndo: false,
  canRedo: false,
  operation: null,
  onSwitchboardLinkClick: undefined,
  selectedTableItem: null,
  selectedTableName: null,
  getIsFormOpen: () => false,
  clearSelected: () => {},
  viewItem: () => {},
  createItem: () => {},
  editItem: () => {},
  onExport: () => {},
  onClose: () => {},
  handleUndo: () => {},
  handleRedo: () => {},
  onShowRevisionHistory: () => {},
  handleAction: () => {},
};

const EditorContext = createContext(defaultEditorContextValue);
EditorContext.displayName = "EditorContext";

type TableState = {
  selectedTableItem: TableItemType<TableName> | null;
  selectedTableName: TableName | null;
  operation: Operation;
};

type TableAction =
  | {
      type: "VIEW_ITEM";
      item: TableItemType<TableName>;
      tableName: TableName;
    }
  | { type: "CREATE_ITEM"; tableName: TableName }
  | {
      type: "EDIT_ITEM";
      item: TableItemType<TableName>;
      tableName: TableName;
    }
  | { type: "CLEAR_SELECTED" };

function tableReducer(state: TableState, action: TableAction): TableState {
  switch (action.type) {
    case "VIEW_ITEM":
      return {
        selectedTableItem: action.item,
        selectedTableName: action.tableName,
        operation: "view",
      };
    case "CREATE_ITEM":
      return {
        selectedTableItem: null,
        selectedTableName: action.tableName,
        operation: "create",
      };
    case "EDIT_ITEM":
      return {
        selectedTableItem: action.item,
        selectedTableName: action.tableName,
        operation: "edit",
      };
    case "CLEAR_SELECTED":
      return {
        selectedTableItem: null,
        selectedTableName: null,
        operation: null,
      };
    default:
      return state;
  }
}

export function RWAEditorContextProvider(
  props: RWAEditorContextProps & {
    readonly children: ReactNode;
    readonly state: RealWorldAssetsState;
    readonly editorDispatcher: EditorDispatcher;
    readonly undo: () => void;
    readonly redo: () => void;
  },
) {
  const {
    children,
    state,
    isAllowedToCreateDocuments,
    isAllowedToEditDocuments,
    canUndo,
    canRedo,
    undo,
    redo,
    editorDispatcher,
    onExport,
    onClose,
    onSwitchboardLinkClick,
    onShowRevisionHistory,
  } = props;
  const [editorState, setEditorState] = useState(state);
  const stateRef = useRef(state);
  const [tableState, tableDispatch] = useReducer(tableReducer, {
    selectedTableItem: null,
    selectedTableName: null,
    operation: null,
  });

  const { selectedTableItem, selectedTableName, operation } = tableState;

  useEffect(() => {
    if (operation === null || operation === "view") {
      setEditorState(stateRef.current);
    }
  }, [operation, state]);

  useEffect(() => {
    stateRef.current = state;
  }, [state]);

  const {
    accounts,
    fixedIncomeTypes,
    portfolio,
    principalLenderAccountId,
    serviceProviderFeeTypes,
    spvs,
    transactions,
  } = editorState;

  const cashAsset = useMemo(() => getCashAsset(portfolio), [portfolio]);

  const fixedIncomes = useMemo(
    () => getFixedIncomeAssets(portfolio),
    [portfolio],
  );

  const handleUndo = useCallback(() => {
    if (!canUndo) return;
    undo();
  }, [canUndo, undo]);

  const handleRedo = useCallback(() => {
    if (!canRedo) return;
    redo();
  }, [canRedo, redo]);

  const getIsFormOpen = useCallback(
    (tableName: TableName) => selectedTableName === tableName,
    [selectedTableName],
  );

  const viewItem = useCallback(
    (item: TableItemType<TableName>, tableName: TableName) => {
      tableDispatch({ type: "VIEW_ITEM", item, tableName });
    },
    [],
  );

  const createItem = useCallback((tableName: TableName) => {
    tableDispatch({ type: "CREATE_ITEM", tableName });
  }, []);

  const editItem = useCallback(
    (item: TableItemType<TableName>, tableName: TableName) => {
      tableDispatch({ type: "EDIT_ITEM", item, tableName });
    },
    [],
  );

  const clearSelected = useCallback(() => {
    tableDispatch({ type: "CLEAR_SELECTED" });
    setEditorState(stateRef.current);
  }, []);

  const handleAction = useCallback(
    (action: EditorAction) => {
      try {
        const result = editorDispatcher(action);
        const actionOperationType = getActionOperationType(action);

        if (actionOperationType === "DELETE" || !result) {
          clearSelected();
          return;
        }
        const tableName = getTableNameFor(action);
        const stateKey = getStateKeyForTableName(tableName);
        const currentStateForKey = stateRef.current[stateKey];
        if (typeof currentStateForKey === "string") return;

        if (actionOperationType === "CREATE") {
          const newStateForKey = [...currentStateForKey, result];
          const newState = {
            ...stateRef.current,
            [stateKey]: newStateForKey,
          };
          stateRef.current = newState;
          setEditorState(newState);

          if (tableName === selectedTableName) {
            const newTableData = makeTableData(tableName, newState);
            const newTableItem = newTableData.find(
              (item) => item.id === result.id,
            );

            if (newTableItem) {
              viewItem(newTableItem, tableName);
            }
          }
        }

        if (actionOperationType === "EDIT") {
          const newStateForKey = currentStateForKey.map((item) =>
            item.id === result.id ? result : item,
          );
          const newState = {
            ...stateRef.current,
            [stateKey]: newStateForKey,
          };
          stateRef.current = newState;
          setEditorState(newState);
          const newTableData = makeTableData(tableName, newState);
          const newTableItem = newTableData.find(
            (item) => item.id === result.id,
          );
          if (newTableItem) {
            editItem(newTableItem, tableName);
          }
        }
      } catch (error) {
        console.error(`Failed to dispatch action ${action.type}`, error);
      }
    },
    [clearSelected, editorDispatcher, editItem, selectedTableName, viewItem],
  );

  const value = useMemo(
    () => ({
      accounts,
      fixedIncomeTypes,
      portfolio,
      principalLenderAccountId,
      serviceProviderFeeTypes,
      spvs,
      transactions,
      cashAsset,
      fixedIncomes,
      canUndo,
      canRedo,
      isAllowedToCreateDocuments,
      isAllowedToEditDocuments,
      selectedTableItem,
      selectedTableName,
      operation,
      handleAction,
      handleUndo,
      handleRedo,
      viewItem,
      createItem,
      editItem,
      clearSelected,
      getIsFormOpen,
      onExport,
      onClose,
      onSwitchboardLinkClick,
      onShowRevisionHistory,
    }),
    [
      accounts,
      canRedo,
      canUndo,
      cashAsset,
      clearSelected,
      createItem,
      editItem,
      fixedIncomeTypes,
      fixedIncomes,
      getIsFormOpen,
      handleAction,
      handleRedo,
      handleUndo,
      isAllowedToCreateDocuments,
      isAllowedToEditDocuments,
      onClose,
      onExport,
      onShowRevisionHistory,
      onSwitchboardLinkClick,
      operation,
      portfolio,
      principalLenderAccountId,
      selectedTableItem,
      selectedTableName,
      serviceProviderFeeTypes,
      spvs,
      transactions,
      viewItem,
    ],
  );
  return (
    <EditorContext.Provider value={value}>{children}</EditorContext.Provider>
  );
}

export function useEditorContext() {
  return useContext(EditorContext);
}
