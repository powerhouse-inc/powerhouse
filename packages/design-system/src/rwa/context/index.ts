export * from "./editor-context.js";
