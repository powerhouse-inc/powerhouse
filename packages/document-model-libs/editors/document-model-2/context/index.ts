export * from "./DocumentModelContext";
export * from "./FormContext";
export * from "./SchemaContext";
export * from "./FormManager";
