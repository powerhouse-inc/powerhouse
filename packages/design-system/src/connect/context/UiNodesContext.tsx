import { DRIVE, FILE, UiDriveNode, UiFolderNode, UiNode } from "@/connect";
import {
  createContext,
  FC,
  ReactNode,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";

export type TUiNodesContext = {
  driveNodes: UiDriveNode[];
  selectedNode: UiNode | null;
  selectedNodePath: UiNode[];
  selectedDriveNode: UiDriveNode | null;
  selectedParentNode: UiDriveNode | UiFolderNode | null;
  setDriveNodes: (driveNodes: UiDriveNode[]) => void;
  setSelectedNode: (node: UiNode | null) => void;
  getNodeById: (id: string) => UiNode | null;
  getParentNode: (uiNode: UiNode) => UiNode | null;
  getIsSelected: (node: UiNode) => boolean;
  getIsInSelectedNodePath: (node: UiNode) => boolean;
  getSiblings: (node: UiNode) => UiNode[];
};

const defaultTreeItemContextValue: TUiNodesContext = {
  driveNodes: [],
  selectedNode: null,
  selectedNodePath: [],
  selectedDriveNode: null,
  selectedParentNode: null,
  setDriveNodes: () => {},
  setSelectedNode: () => {},
  getNodeById: () => null,
  getParentNode: () => null,
  getIsSelected: () => false,
  getIsInSelectedNodePath: () => false,
  getSiblings: () => [],
};

export const UiNodesContext = createContext<TUiNodesContext>(
  defaultTreeItemContextValue,
);

export interface UiNodesContextProviderProps {
  readonly children: ReactNode;
}

export const UiNodesContextProvider: FC<UiNodesContextProviderProps> = ({
  children,
}) => {
  const [driveNodes, setDriveNodes] = useState<UiDriveNode[]>([]);
  const [selectedNode, _setSelectedNode] = useState<UiNode | null>(null);
  const [selectedNodePath, setSelectedNodePath] = useState<UiNode[]>([]);
  const [selectedDriveNode, setSelectedDriveNode] =
    useState<UiDriveNode | null>(null);
  const [selectedParentNode, setSelectedParentNode] = useState<
    UiDriveNode | UiFolderNode | null
  >(null);

  /*
        This internal function that uses `driveNodes` as an argument to prevent stale objects being used in the closure.
        External `getNodeById` for use by other components doesn't need to do this because it is only invoked by external functions, and can therefore omit this argument for convenience.
     */
  const _getNodeById = useCallback(
    (id: string, driveNodes: UiDriveNode[] | null) => {
      if (!driveNodes?.length) return null;

      for (const driveNode of driveNodes) {
        if (driveNode.id === id) return driveNode;

        const node = driveNode.nodeMap[id];

        if (node) return node;
      }

      return null;
    },
    [],
  );

  const getNodeById = useCallback(
    (id: string) => {
      return _getNodeById(id, driveNodes);
    },
    [_getNodeById, driveNodes],
  );

  const getSelectedDriveNode = useCallback(
    (selectedNode: UiNode | null, driveNodes: UiDriveNode[] | null) => {
      if (!selectedNode || !driveNodes?.length) return null;

      if (selectedNode.kind === DRIVE) return selectedNode;

      return driveNodes.find((d) => d.id === selectedNode.driveId) ?? null;
    },
    [],
  );

  /*
        This internal function that uses `driveNodes` as an argument to prevent stale objects being used in the closure.
        External `getNodeById` for use by other components doesn't need to do this because it is only invoked by external functions, and can therefore omit this argument for convenience.
     */
  const _getParentNode = useCallback(
    (node: UiNode, driveNodes: UiDriveNode[] | null) => {
      if (!driveNodes?.length || node.kind === DRIVE) return null;

      const parentNode = _getNodeById(node.parentFolder, driveNodes);

      if (!parentNode) return null;

      if (parentNode.kind === FILE) {
        throw new Error(
          `Parent node ${node.parentFolder} is a file, not a folder`,
        );
      }

      return parentNode;
    },
    [_getNodeById],
  );

  const getParentNode = useCallback(
    (uiNode: UiNode) => {
      return _getParentNode(uiNode, driveNodes);
    },
    [_getParentNode, driveNodes],
  );

  const getSelectedParentNode = useCallback(
    (selectedNode: UiNode | null, driveNodes: UiDriveNode[] | null) => {
      if (!selectedNode || !driveNodes?.length) return null;

      if (selectedNode.kind === FILE)
        return _getParentNode(selectedNode, driveNodes);

      return selectedNode;
    },
    [_getParentNode],
  );

  const getPathToNode = useCallback(
    (uiNode: UiNode, driveNodes: UiDriveNode[]) => {
      const path: UiNode[] = [];

      const driveNode = driveNodes.find((d) => d.id === uiNode.driveId);

      let current: UiNode | undefined = uiNode;

      while (current) {
        path.push(current);
        current =
          current.parentFolder === driveNode?.id
            ? driveNode
            : current.parentFolder
              ? driveNode?.nodeMap[current.parentFolder]
              : undefined;
      }

      return path.reverse();
    },
    [],
  );

  /*
     _setSelectedNode from `useState` is kept internal so that we can instead expose this function, which also manages the selectedDriveNode, selectedParentNode, and selectedNodePath states.
    */
  const setSelectedNode = useCallback(
    (uiNode: UiNode | null) => {
      if (!uiNode) return;

      _setSelectedNode(uiNode);
      setSelectedDriveNode(getSelectedDriveNode(uiNode, driveNodes));
      setSelectedParentNode(getSelectedParentNode(uiNode, driveNodes));

      if (uiNode.kind === DRIVE) {
        setSelectedNodePath([uiNode]);
        return;
      }

      const newSelectedNodePath = getPathToNode(uiNode, driveNodes);

      setSelectedNodePath(newSelectedNodePath);
    },
    [driveNodes, getPathToNode, getSelectedDriveNode, getSelectedParentNode],
  );

  const getIsSelected = useCallback(
    (node: UiNode) => {
      return selectedNode === node;
    },
    [selectedNode],
  );

  const getIsInSelectedNodePath = useCallback(
    (node: UiNode) => {
      if (node.kind === FILE) return false;
      return selectedNodePath.includes(node);
    },
    [selectedNodePath],
  );

  const getSiblings = useCallback(
    (node: UiNode) => {
      if (node.kind === DRIVE) {
        console.warn(
          "Drive nodes do not have siblings, as they are top-level nodes",
        );
        return [];
      }

      const parent = _getParentNode(node, driveNodes);

      return parent?.children ?? [];
    },
    [_getParentNode, driveNodes],
  );

  useEffect(() => {
    if (!selectedNode) return;

    const updatedSelectedNode = _getNodeById(selectedNode.id, driveNodes);

    if (updatedSelectedNode) {
      setSelectedNode(updatedSelectedNode);
    }
  }, [driveNodes, _getNodeById, selectedNode, setSelectedNode]);

  const value = useMemo(
    () => ({
      driveNodes,
      selectedNode,
      selectedNodePath,
      selectedDriveNode,
      selectedParentNode,
      getNodeById,
      getParentNode,
      setDriveNodes,
      setSelectedNode,
      getIsSelected,
      getIsInSelectedNodePath,
      getSiblings,
    }),
    [
      driveNodes,
      selectedNode,
      selectedNodePath,
      selectedDriveNode,
      selectedParentNode,
      getNodeById,
      getParentNode,
      setSelectedNode,
      getIsSelected,
      getIsInSelectedNodePath,
      getSiblings,
    ],
  );

  return (
    <UiNodesContext.Provider value={value}>{children}</UiNodesContext.Provider>
  );
};

export const useUiNodesContext = () => {
  const contextValue = useContext(UiNodesContext);
  return contextValue;
};
