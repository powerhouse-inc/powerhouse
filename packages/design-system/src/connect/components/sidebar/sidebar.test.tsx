import {
  render,
  screen,
  waitForElementToBeRemoved,
} from "@testing-library/react";

import { WagmiContext } from "@/connect/context/WagmiContext";
import { ConnectSidebar } from ".";

describe("Connect Sidebar Component", () => {
  it("should match snapshot", async () => {
    const { asFragment } = render(
      <WagmiContext>
        <ConnectSidebar
          address="0x123"
          collapsed={false}
          data-testid="sidebar"
          onLogin={() => {}}
          onToggle={() => {}}
        />
      </WagmiContext>,
    );

    await waitForElementToBeRemoved(() =>
      screen.getAllByTestId("icon-fallback"),
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render correctly", () => {
    render(
      <WagmiContext>
        <ConnectSidebar
          address="0x123"
          collapsed={false}
          data-testid="sidebar"
          onLogin={() => {}}
          onToggle={() => {}}
        />
      </WagmiContext>,
    );
    expect(screen.getByTestId("sidebar")).toBeInTheDocument();
  });

  it("should render expanded", () => {
    render(
      <WagmiContext>
        <ConnectSidebar
          address="0x123"
          collapsed={false}
          data-testid="sidebar"
          onLogin={() => {}}
          onToggle={() => {}}
        />
      </WagmiContext>,
    );
    expect(screen.getByTestId("sidebar")).toHaveStyle({ width: "304px" });
  });

  it("should render collapsed", () => {
    render(
      <WagmiContext>
        <ConnectSidebar
          address="0x123"
          collapsed
          data-testid="sidebar"
          onLogin={() => {}}
          onToggle={() => {}}
        />
      </WagmiContext>,
    );
    expect(screen.getByTestId("sidebar")).toHaveStyle({ width: "58px" });
  });
});
