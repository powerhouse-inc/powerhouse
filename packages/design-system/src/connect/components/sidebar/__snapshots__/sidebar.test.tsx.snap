// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Connect Sidebar Component > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="group flex h-full flex-col bg-slate-50"
    data-testid="sidebar"
    style="width: 58px;"
  >
    <div
      class="no-scrollbar flex-1 overflow-auto text-gray-900 transition-shadow"
    >
      <div
        class="shrink-0 flex justify-center gap-4 border-b border-gray-300 py-4"
      >
        <button
          class=""
          type="button"
        >
          <svg
            class="text-gray-600"
            fill="none"
            height="24"
            style="width: 24px; height: 24px;"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11.0306 3.68615C11.2854 3.68614 11.492 3.47932 11.4921 3.2242L11.4921 9.03413e-06C8.29133 0.132881 5.41451 1.52048 3.33928 3.68633L11.0306 3.68615ZM14.2609 8.42821C14.2609 8.68334 14.0543 8.89016 13.7994 8.89016L0.406086 8.89043C0.81778 7.34998 1.52777 5.93103 2.46931 4.70043L13.7995 4.70024C14.0544 4.70023 14.261 4.90707 14.261 5.16221L14.2609 8.42821ZM20.6591 20.3137C18.5847 22.4787 15.7079 23.8663 12.5062 24L12.5063 20.7758C12.5063 20.5207 12.7129 20.3139 12.9678 20.3139L20.6591 20.3137ZM11.4924 23.9992C8.29157 23.8665 5.41401 22.4782 3.33969 20.3133L11.0309 20.3131C11.2858 20.3131 11.4924 20.52 11.4924 20.7751L11.4924 23.9992ZM17.5946 15.5703C17.5947 15.3152 17.8013 15.1083 18.0561 15.1083L23.5923 15.108C23.1806 16.6484 22.4714 18.0666 21.5291 19.298L18.0561 19.2983C17.8012 19.2983 17.5946 19.0914 17.5946 18.8363L17.5946 15.5703ZM8.84765 10.3544C8.85393 10.1039 9.05863 9.90407 9.30899 9.90407L23.8175 9.90396C23.9373 10.584 24 11.2842 24 11.9989C24 12.7136 23.9372 13.4138 23.8174 14.0939L9.31089 14.0933C9.06005 14.0932 8.85517 13.8927 8.84957 13.6416L8.80986 11.8617L8.84765 10.3544ZM16.7389 18.8372C16.7389 19.0923 16.5323 19.2991 16.2774 19.2991L2.46978 19.2992C1.52749 18.0679 0.818373 16.6498 0.406752 15.1093L16.2775 15.1092C16.5324 15.1092 16.739 15.316 16.739 15.5712L16.7389 18.8372ZM0.182552 9.90452L7.3636 9.90412C7.61849 9.90411 7.82511 10.1109 7.82511 10.3661L7.82503 13.6321C7.82503 13.8872 7.61842 14.0941 7.36355 14.0941L0.182447 14.0945C0.062664 13.4144 -1.71661e-05 12.7142 0 11.9995C1.71661e-05 11.2848 0.0627384 10.5846 0.182552 9.90452ZM15.2756 5.16214C15.2757 4.90701 15.4823 4.70019 15.7371 4.70019L21.5294 4.69998C22.4717 5.93135 23.1808 7.34943 23.5924 8.88987L15.7371 8.89013C15.4822 8.89014 15.2756 8.6833 15.2756 8.42816L15.2756 5.16214ZM12.5068 0C15.7076 0.132721 18.5852 1.52098 20.6595 3.68592L12.9682 3.6861C12.7133 3.6861 12.5067 3.47927 12.5067 3.22413L12.5068 0Z"
              fill="#9EA0A1"
            />
          </svg>
        </button>
      </div>
      <div
        class="flex flex-col"
      />
    </div>
    <div
      class="shrink-0 flex flex-col gap-2 border-t border-gray-300 px-2 py-4"
    >
      <div
        class=""
      >
        <div
          class="flex items-center justify-center rounded-sm"
        >
          <div
            class="fade-out flex-none animate-pulse rounded-full bg-gray-400"
            style="width: 40px; height: 40px;"
          />
        </div>
      </div>
      <button
        class="mt-3 flex w-full cursor-pointer items-center justify-center outline-none"
      >
        <svg
          class="text-gray-600"
          fill="currentcolor"
          style="width: 24px; height: 24px;"
          viewBox="0 0 24 24"
        >
          <path
            clip-rule="evenodd"
            d="M12.5 2C13.1894 2 13.8136 2.27905 14.2659 2.73036C14.2834 2.74784 14.3007 2.76559 14.3177 2.78358C14.7174 3.20678 14.971 3.76953 14.9977 4.39107L14.9985 4.41205C14.9995 4.44124 15 4.47056 15 4.5L15 4.51426C15 4.55491 15.0247 4.59145 15.0623 4.60702C15.0998 4.62261 15.1431 4.61423 15.1719 4.58546L15.182 4.57538C15.2028 4.55456 15.2239 4.53419 15.2452 4.51426L15.2607 4.5C15.719 4.07934 16.2962 3.86073 16.8782 3.84417C16.9029 3.84346 16.9277 3.84312 16.9524 3.84315C17.5913 3.84382 18.23 4.0879 18.7175 4.57538L19.4246 5.28249C19.9121 5.76997 20.1562 6.40868 20.1569 7.0476C20.1569 7.07234 20.1565 7.09709 20.1558 7.12182C20.1393 7.70375 19.9207 8.281 19.5 8.73935L19.4857 8.75476C19.4658 8.77611 19.4454 8.7972 19.4246 8.81802L19.4145 8.82811C19.3858 8.85687 19.3774 8.90017 19.393 8.93775C19.4085 8.97529 19.4451 9 19.4857 9L19.5 9C19.5294 9 19.5588 9.00051 19.5879 9.00152L19.6089 9.00233C20.2305 9.02898 20.7932 9.28257 21.2164 9.68233C21.2344 9.69933 21.2522 9.71659 21.2696 9.73411C21.721 10.1864 22 10.8106 22 11.5L22 12.5C22 13.1894 21.721 13.8136 21.2696 14.2659C21.2522 14.2834 21.2344 14.3007 21.2164 14.3177C20.7932 14.7174 20.2305 14.971 19.6089 14.9977L19.5879 14.9985C19.5588 14.9995 19.5294 15 19.5 15L19.4857 15C19.4451 15 19.4085 15.0247 19.393 15.0623C19.3774 15.0998 19.3858 15.1431 19.4145 15.1719L19.4246 15.182C19.4454 15.2028 19.4658 15.2239 19.4857 15.2452L19.5 15.2607C19.9207 15.719 20.1393 16.2962 20.1558 16.8782C20.1565 16.9029 20.1569 16.9277 20.1569 16.9524C20.1562 17.5913 19.9121 18.23 19.4246 18.7175L18.7175 19.4246C18.23 19.9121 17.5913 20.1562 16.9524 20.1568C16.9277 20.1569 16.9029 20.1565 16.8782 20.1558C16.2963 20.1393 15.719 19.9207 15.2607 19.5L15.2452 19.4857C15.2239 19.4658 15.2028 19.4454 15.182 19.4246L15.1719 19.4145C15.1431 19.3858 15.0998 19.3774 15.0623 19.393C15.0247 19.4085 15 19.4451 15 19.4857L15 19.5C15 19.5294 14.9995 19.5588 14.9985 19.5879L14.9977 19.6089C14.971 20.2305 14.7174 20.7932 14.3177 21.2164C14.3007 21.2344 14.2834 21.2522 14.2659 21.2696C13.8136 21.7209 13.1894 22 12.5 22L11.5 22C10.8106 22 10.1864 21.7209 9.73411 21.2696C9.71659 21.2522 9.69933 21.2344 9.68233 21.2164C9.28257 20.7932 9.02898 20.2305 9.00233 19.6089L9.00152 19.5879C9.00051 19.5588 9 19.5294 9 19.5L9 19.4857C9 19.4451 8.97529 19.4085 8.93774 19.393C8.90017 19.3774 8.85687 19.3858 8.8281 19.4145L8.81802 19.4246C8.7972 19.4454 8.77611 19.4658 8.75475 19.4857L8.73934 19.5C8.281 19.9207 7.70375 20.1393 7.12182 20.1558C7.09708 20.1565 7.07234 20.1569 7.0476 20.1568C6.40868 20.1562 5.76997 19.9121 5.28249 19.4246L4.57538 18.7175C4.0879 18.23 3.84382 17.5913 3.84315 16.9524C3.84312 16.9277 3.84346 16.9029 3.84417 16.8782C3.86073 16.2962 4.07934 15.719 4.5 15.2607L4.51426 15.2452C4.53419 15.2239 4.55456 15.2028 4.57538 15.182L4.58546 15.1719C4.61423 15.1431 4.62261 15.0998 4.60702 15.0623C4.59145 15.0247 4.55491 15 4.51426 15L4.5 15C4.47056 15 4.44125 14.9995 4.41205 14.9985L4.39107 14.9977C3.76953 14.971 3.20678 14.7174 2.78358 14.3177C2.76558 14.3007 2.74784 14.2834 2.73036 14.2659C2.27905 13.8136 2 13.1894 2 12.5L2 11.5C2 10.8106 2.27905 10.1864 2.73036 9.73411C2.74784 9.71659 2.76558 9.69933 2.78358 9.68233C3.20678 9.28257 3.76953 9.02898 4.39107 9.00233L4.41205 9.00152C4.44125 9.00051 4.47056 9 4.5 9L4.51426 9C4.55491 9 4.59145 8.97529 4.60702 8.93775C4.62261 8.90017 4.61423 8.85687 4.58546 8.82811L4.57538 8.81802C4.55456 8.7972 4.53419 8.77611 4.51426 8.75476L4.5 8.73935C4.07934 8.281 3.86073 7.70375 3.84417 7.12182C3.84346 7.09709 3.84312 7.07234 3.84315 7.0476C3.84382 6.40868 4.0879 5.76997 4.57538 5.28249L5.28249 4.57538C5.76997 4.0879 6.40868 3.84382 7.0476 3.84315C7.07234 3.84312 7.09709 3.84346 7.12182 3.84417C7.70375 3.86073 8.281 4.07934 8.73934 4.5L8.75476 4.51426C8.77611 4.53419 8.7972 4.55456 8.81802 4.57538L8.8281 4.58546C8.85687 4.61423 8.90017 4.62261 8.93775 4.60702C8.97529 4.59145 9 4.55491 9 4.51426L9 4.5C9 4.47056 9.00051 4.44124 9.00152 4.41205L9.00233 4.39107C9.02898 3.76953 9.28257 3.20678 9.68233 2.78358C9.69933 2.76559 9.71659 2.74784 9.73411 2.73036C10.1864 2.27905 10.8106 2 11.5 2L12.5 2ZM11 19.5C11 19.7761 11.2239 20 11.5 20L12.5 20C12.7761 20 13 19.7761 13 19.5L13 19.4857C13 18.6262 13.5212 17.8669 14.2962 17.5455C15.0727 17.2235 15.9787 17.3929 16.5861 18.0003L16.5962 18.0104C16.7915 18.2057 17.108 18.2057 17.3033 18.0104L18.0104 17.3033C18.2057 17.108 18.2057 16.7915 18.0104 16.5962L18.0003 16.5861C17.3929 15.9787 17.2235 15.0727 17.5455 14.2961C17.8669 13.5212 18.6262 13 19.4857 13L19.5 13C19.7761 13 20 12.7761 20 12.5L20 11.5C20 11.2239 19.7761 11 19.5 11L19.4857 11C18.6262 11 17.8669 10.4788 17.5455 9.70385C17.2235 8.92727 17.3929 8.02132 18.0003 7.41389L18.0104 7.40381C18.2057 7.20854 18.2057 6.89196 18.0104 6.6967L17.3033 5.98959C17.108 5.79433 16.7915 5.79433 16.5962 5.98959L16.5861 5.99968C15.9787 6.60711 15.0727 6.77651 14.2962 6.45448C13.5212 6.13311 13 5.37381 13 4.51426L13 4.5C13 4.22386 12.7761 4 12.5 4L11.5 4C11.2239 4 11 4.22386 11 4.5L11 4.51426C11 5.37381 10.4788 6.13311 9.70384 6.45448C8.92725 6.77651 8.02132 6.60711 7.41389 5.99968L7.40381 5.98959C7.20854 5.79433 6.89196 5.79433 6.6967 5.98959L5.98959 6.6967C5.79433 6.89196 5.79433 7.20854 5.98959 7.40381L5.99967 7.41389C6.60711 8.02132 6.77651 8.92727 6.45448 9.70385C6.13311 10.4788 5.37382 11 4.51426 11L4.5 11C4.22386 11 4 11.2239 4 11.5L4 12.5C4 12.7761 4.22386 13 4.5 13L4.51426 13C5.37382 13 6.13311 13.5212 6.45447 14.2961C6.77651 15.0727 6.60711 15.9787 5.99968 16.5861L5.98959 16.5962C5.79433 16.7915 5.79433 17.108 5.98959 17.3033L6.6967 18.0104C6.89196 18.2057 7.20854 18.2057 7.4038 18.0104L7.41389 18.0003C8.02132 17.3929 8.92726 17.2235 9.70384 17.5455C10.4788 17.8669 11 18.6262 11 19.4857L11 19.5Z"
            fill-rule="evenodd"
          />
          <path
            clip-rule="evenodd"
            d="M12 14C13.1046 14 14 13.1046 14 12C14 10.8954 13.1046 10 12 10C10.8954 10 10 10.8954 10 12C10 13.1046 10.8954 14 12 14ZM12 16C14.2091 16 16 14.2091 16 12C16 9.79086 14.2091 8 12 8C9.79086 8 8 9.79086 8 12C8 14.2091 9.79086 16 12 16Z"
            fill-rule="evenodd"
          />
        </svg>
        <span
          class="hidden text-sm font-semibold leading-6 text-gray-800"
        >
          Settings
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
